lib/drives/models/drive.dart

```diff
@@ -106,9 +106,43 @@ class Drive extends Trip {
     return maxUsedSeats;
   }
 
+  bool isRidePossible(Ride ride) {
+    List<Ride> consideredRides = approvedRides! + [ride];
+    Set<DateTime> times = consideredRides
+        .map((consideredRide) => [consideredRide.startTime, consideredRide.endTime])
+        .expand((x) => x)
+        .toSet();
+    for (DateTime time in times) {
+      int usedSeats = 0;
+      for (Ride consideredRide in consideredRides) {
+        final startTimeBeforeOrEqual =
+            consideredRide.startTime.isBefore(time) || consideredRide.startTime.isAtSameMomentAs(time);
+        final endTimeAfter = consideredRide.endTime.isAfter(time);
+        if (startTimeBeforeOrEqual && endTimeAfter) {
+          usedSeats += consideredRide.seats;
+        }
+      }
+
+      if (usedSeats > seats) {
+        return false;
+      }
+    }
+    return true;
+  }
+
   Future<void> cancel() async {
     cancelled = true;
     await supabaseClient.from('drives').update({'cancelled': true}).eq('id', id);
-    await supabaseClient.from('rides').update({'status': RideStatus.cancelledByDriver.index}).eq('drive_id', id);
+    //cancel all pending and approved rides for this drive.
+    await supabaseClient
+        .from('rides')
+        .update({'status': RideStatus.cancelledByDriver.index})
+        .eq('drive_id', id)
+        .or('status.eq.${RideStatus.pending.index},status.eq.${RideStatus.approved.index}');
```

Comment:
I am wondering if this can be controlled on the supabase side?

i.e. that a ride cannot change status from cancelledByRider to cancelledByDriver

Reply:
yes i'm working on that but i'll leave it for now